truffle/src/main/java/org/truffleruby/core/queue/UnsizedQueue.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (c) 2017 Oracle and/or its affiliates. All rights reserved. This
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 1.0
+ * GNU General Public License version 2
+ * GNU Lesser General Public License version 2.1
+ */
+package org.truffleruby.core.queue;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class UnsizedQueue {
+
+    private final ReentrantLock lock = new ReentrantLock();
+    private final Condition canTake = lock.newCondition();
+
+    private Item addEnd;
+    private Item takeEnd;
+    private int size;
+
+    public void add(Object item) {
+        lock.lock();
+
+        try {
+            final Item newItem = new Item(item);
+            if (addEnd != null) {
+                addEnd.setNextToTake(newItem);
+            }
+            addEnd = newItem;
+            if (takeEnd == null) {
+                takeEnd = addEnd;
+            }
+            size++;
+            canTake.signal();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public Object poll() {
+        lock.lock();
+
+        try {
+            if (takeEnd == null) {
+                return null;
+            } else {
+                return doTake();
+            }
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public Object take() throws InterruptedException {
+        lock.lock();
+
+        try {
+            while (takeEnd == null) {
+                canTake.await();
+            }
+
+            return doTake();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public Object poll(long timeout, TimeUnit unit) throws InterruptedException {
+        lock.lock();
+
+        try {
+            while (takeEnd == null) {
+                // We'll repeat the same timeout period again if we skip in this loop, but it's not a guarantee of anyway
+                canTake.await(timeout, unit);
```

Comment:
Could you convert to some small unit and deduce what you waited? Otherwise random InterruptException might make this run forever.